### PR TITLE
AdaptiveAvgPool: derivatives and tests

### PR DESCRIPTION
**Pull request by schaefertim**
Introduces derivatives for AdaptiveAvgPool.
AdaptiveAvgPoolNDDerivatives has a method check_parameters() that computes the equivalent parameters for AvgPool. This method is tested separately.
The input-output-derivatives are tested in the usual test suite.

### CODE REVIEW

**Comment by f-dangel** [comment]
Referencing https://github.com/fKunstner/backpack-discuss/issues/28 for documentation

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # check length of input shape
+        if not len(shape_input) == (self.N + 2):
+            raise NotImplementedError(
+                f"input must be (batch_size, C, ...) with ... {self.N} dimensions"
+            )
+        # check length of target shape
+        if not len(shape_target) == (self.N + 2):
+            raise NotImplementedError(
+                f"shape of target should have {self.N} dimensions"
+            )
```
Comment (schaefertim) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:49]: Is this check necessary?

Comment (f-dangel) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:49]: Do you mean that violating inputs would be caught by the forward pass of `AvgPoolNd`?

Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:49]: Exactly. And consequently, outputs should be fine as well.

**Comment by schaefertim** [comment]
Computing derivatives on AdaptiveAvgPool3d fails on cuda!!! (but works on cpu).
I have no idea what's happening there. This is a very dangerous error, because it doesn't show up in the github test suite.
run `pytest -vx test/core/derivatives/ -k "AdaptiveAvgPool3d"` to reproduce

**Comment by f-dangel** [comment]
Can you post the error message?

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+class AdaptiveAvgPoolNDDerivatives(AvgPoolNDDerivatives):
+    """Implements the derivatives for AdaptiveAvgPool."""
+
+    def __init__(self, N: int):
+        """Initialization.
+
+        Args:
+            N: number of dimensions
+        """
+        super(AdaptiveAvgPoolNDDerivatives, self).__init__(N=N)
```
Comment (f-dangel) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:19]: This is only calling the parent class constructor -> remove?

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 1 above)
Comment (f-dangel) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:49]: I think it is sufficient to only check the input shape and assume the forward pass preserves the dimensions.

I vote for keeping the input check, as adaptive average pooling also works on unbatched inputs:
```python
from torch import rand
from torch.nn.functional import adaptive_avg_pool2d

# unbatched image
input = rand(3, 20, 20)
output = adaptive_avg_pool2d(input, output_size=(3, 3))
print(output.shape) # -> torch.Size([3, 3, 3])

# batched image
input = rand(1, 3, 20, 20)
output = adaptive_avg_pool2d(input, output_size=(3, 3))
print(output.shape) # -> torch.Size([1, 3, 3, 3])
```
The derivatives assume there *is a batch dimension*, even if `batch_size = 1`.

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        # check if input shape is multiple of output shape
+        for n in range(self.N):
+            in_dim: int = shape_input[2 + n]
+            out_dim: int = shape_target[2 + n]
+            if (in_dim % out_dim) != 0:
+                raise NotImplementedError(
+                    "Not supported in BackPACK: Input shape of AdaptiveAvgPool "
+                    "must be multiple of output shape. Only in this case, there is "
+                    "an equivalent AvgPool module."
```
Comment (f-dangel) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:59]: Shorter suggestion with more feedback: `f"No equivalent unadaptive pooling: Input shape ({shape_input}) must be multiple of output shape ({shape_output})."`

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        if not len(shape_target) == (self.N + 2):
+            raise NotImplementedError(
+                f"shape of target should have {self.N} dimensions"
+            )
+
+        # check if input shape is multiple of output shape
+        for n in range(self.N):
+            in_dim: int = shape_input[2 + n]
+            out_dim: int = shape_target[2 + n]
+            if (in_dim % out_dim) != 0:
```
Comment (f-dangel) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:55]: Shorter suggestion:
```
if any(in_dim % out_dim != 0 for (in_dim, out_dim) in zip(shape_input[2:], shape_target[2:])):
   ...
```

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        # calculate equivalent AvgPoolND parameters
+        stride: List[int] = []
+        kernel_size: List[int] = []
+        for n in range(self.N):
+            in_dim: int = shape_input[2 + n]
+            out_dim: int = shape_target[2 + n]
+            stride.append(in_dim // out_dim)
+            kernel_size.append(in_dim - (out_dim - 1) * stride[n])
+
```
Comment (f-dangel) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:87]: - Explicitly add `padding = [0 for _ in range(N)]`
- Adapt `return ..., ..., padding`

File: test/core/derivatives/pooling_adaptive_settings.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+###############################################################################
+
+example = {
+    "module_fn": lambda: torch.nn.AdaptiveAvgPool2d(output_size=(2, 2)),
+    "input_fn": lambda: torch.rand(size=(1, 3, 8, 8)),
+    "device": [torch.device("cpu")],  # optional
+    "seed": 0,  # optional
+    "id_prefix": "pooling-example",  # optional
+}
+POOLING_ADAPTIVE_SETTINGS.append(example)
```
Comment (f-dangel) [inline test/core/derivatives/pooling_adaptive_settings.py:31]: remove?

File: test/adaptive_avg_pool/test_adaptive_avg_pool_nd.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        problem.check_parameters()
+        problem.check_equivalence()
+    else:
+        with pytest.raises(NotImplementedError):
+            problem.check_parameters()
+        # TODO improve equivalence coverage and activate these lines
+        # check for perfect coverage:
+        # all cases that have equivalent parameters are recognized
+        # with pytest.raises(AssertionError):
+        #    problem.check_equivalence()
```
Comment (f-dangel) [inline test/adaptive_avg_pool/test_adaptive_avg_pool_nd.py:32]: Is this still a TODO?

File: test/adaptive_avg_pool/problem.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    required = ["N", "shape_input", "shape_target", "works"]
+    optional = {
+        "id_prefix": "",
+        "device": get_available_devices(),
+        "seed": 0,
+    }
+
+    for req in required:
+        if req not in setting.keys():
+            raise ValueError("Missing configuration entry for {}".format(req))
```
Comment (f-dangel) [inline test/adaptive_avg_pool/problem.py:69]: Replace `.format` with `f"..."`

File: test/adaptive_avg_pool/problem.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+    for opt, default in optional.items():
+        if opt not in setting.keys():
+            setting[opt] = default
+
+    for s in setting.keys():
+        if s not in required and s not in optional.keys():
+            raise ValueError("Unknown config: {}".format(s))
```
Comment (f-dangel) [inline test/adaptive_avg_pool/problem.py:77]: Replace `.format` with `f"..."`

File: test/adaptive_avg_pool/problem.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+    def _make_module(self) -> Union[AdaptiveAvgPool1d, AdaptiveAvgPool2d]:
+        if self.N == 1:
+            module = AdaptiveAvgPool1d(output_size=self.shape_target)
+        elif self.N == 2:
+            module = AdaptiveAvgPool2d(output_size=self.shape_target)
+        elif self.N == 3:
+            module = AdaptiveAvgPool3d(output_size=self.shape_target)
+        else:
+            raise NotImplementedError(f"N={self.N} not implemented in test suite.")
```
Comment (f-dangel) [inline test/adaptive_avg_pool/problem.py:147]: move check to constructor

File: test/adaptive_avg_pool/problem.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        self, stride: List[int], kernel_size: List[int]
+    ) -> Union[AvgPool1d, AvgPool2d]:
+        if self.N == 1:
+            module = AvgPool1d(kernel_size=kernel_size[0], stride=stride[0])
+        elif self.N == 2:
+            module = AvgPool2d(kernel_size=kernel_size, stride=stride)
+        elif self.N == 3:
+            module = AvgPool3d(kernel_size=kernel_size, stride=stride)
+        else:
+            raise NotImplementedError(f"N={self.N} not implemented in test suite.")
```
Comment (f-dangel) [inline test/adaptive_avg_pool/problem.py:185]: move check to constructor

**Comment by schaefertim** [comment]
> Can you post the error message?

Tests:
```
test/core/derivatives/derivatives_test.py::test_jac_mat_prod[dev=cpu-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] PASSED [  8%]
test/core/derivatives/derivatives_test.py::test_jac_mat_prod[dev=cuda-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] PASSED [ 16%]
test/core/derivatives/derivatives_test.py::test_jac_t_mat_prod[dev=cpu-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] PASSED [ 25%]
test/core/derivatives/derivatives_test.py::test_jac_t_mat_prod[dev=cuda-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] FAILED [ 33%]
test/core/derivatives/derivatives_test.py::test_ea_jac_t_mat_jac_prod[dev=cpu-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] PASSED [ 41%]
test/core/derivatives/derivatives_test.py::test_ea_jac_t_mat_jac_prod[dev=cuda-in=(1,3,4,8,8)-AdaptiveAvgPool3d(output_size=(2,2,2))] FAILED [ 50%]
```
Error:
`test/automated_test.py:91: AssertionError`

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 1 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:49]: Nice test. I deleted output check and keep input check.

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 2 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:19]: Removed.

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 3 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:59]: Changed it to `f"No equivalent AvgPool (unadaptive): Input shape ({shape_input}) must be multiple of output shape ({shape_output})."`

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 4 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:55]: Changed it to `if any([shape_input[2 + n] % shape_output[2 + n] != 0 for n in range(self.N)]):`

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 5 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:87]: Yes.

File: test/core/derivatives/pooling_adaptive_settings.py (hunk 6 above)
Comment (schaefertim) [reply test/core/derivatives/pooling_adaptive_settings.py:31]: Yes.

File: test/adaptive_avg_pool/test_adaptive_avg_pool_nd.py (hunk 7 above)
Comment (schaefertim) [reply test/adaptive_avg_pool/test_adaptive_avg_pool_nd.py:32]: Yes. Depending on perspective this is either a minor or a major issue.
Probably, I need to improve the explanation, but I'm not sure how to state it concisely.

If `check_parameters` decides there is equivalence, then the equivalence works. (core functionality)
If `check_parameters` decides there is no equivalence, then there are still some cases where the equivalence works.

It would be nice, if all cases that have equivalence are recognized by `check_parameters`. If this was the case, we could activate these lines and the test would still pass (which it doesn't at the moment).

File: test/adaptive_avg_pool/problem.py (hunk 8 above)
Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py:69]: Yes.

File: test/adaptive_avg_pool/problem.py (hunk 9 above)
Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py:77]: Yes.

File: test/adaptive_avg_pool/problem.py (hunk 10 above)
Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py:147]: Yes.

File: test/adaptive_avg_pool/problem.py (hunk 11 above)
Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py:185]: Yes. However, I am leaving a simple `raise NotImplementedError` because otherwise checkers are complaining that `Local variable 'module' might be referenced before assignment`

**Comment by f-dangel** [comment]
Is there a traceback for the `AssertionError`?

**Comment by schaefertim** [comment]
> Is there a traceback for the `AssertionError`?

It's the usual derivatives don't match.
It is no error in the computation. Just the results between torch and backpack don't match on CUDA.
Next test is going to be to compare torch cpu with torch cuda.

File: test/adaptive_avg_pool/problem.py (hunk 11 above)
Comment (f-dangel) [reply test/adaptive_avg_pool/problem.py]: How about
```
pooling_cls = {1: AvgPool, ...}
module = pooling_cls[self.N](kernel_size, ...)
```
which avoids the assignment problem?

Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py]: Yes. (I never used this pattern but makes a lot of sense)

**Comment by schaefertim** [comment]
I tracked down the issue in PyTorch of `AdaptiveAvgPool3d`:
Forward pass is identical for cpu and cuda.
`torch.autograd.grad` is different between cpu and cuda.
(BackPACK does match the cpu derivative in both cases.)

How to proceed?

**Comment by f-dangel** [comment]
Agree. I would swap 2. and 3., and include a link to the issue in the warning/exception.

File: test/core/derivatives/pooling_adaptive_settings.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        "module_fn": lambda: torch.nn.AdaptiveAvgPool1d(output_size=(3,)),
+        "input_fn": lambda: torch.rand(size=(1, 4, 9)),
+    },
+    {
+        "module_fn": lambda: torch.nn.AdaptiveAvgPool2d(output_size=(3, 5)),
+        "input_fn": lambda: torch.rand(size=(1, 3, 9, 20)),
+    },
+    {
+        "module_fn": lambda: torch.nn.AdaptiveAvgPool3d(output_size=(2, 2, 2)),
+        "input_fn": lambda: torch.rand(size=(1, 3, 4, 8, 8)),
```
Comment (f-dangel) [inline test/core/derivatives/pooling_adaptive_settings.py:34]: Can you change one case to `batch_size>1`?

File: backpack/core/derivatives/adaptive_avg_pool_nd.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        if any([shape_input[2 + n] % shape_output[2 + n] != 0 for n in range(self.N)]):
+            raise NotImplementedError(
+                f"No equivalent AvgPool (unadaptive): Input shape ({shape_input}) "
+                f"must be multiple of output shape ({shape_output})."
+            )
+
+    def get_parameters(
+        self, module: Union[AdaptiveAvgPool1d, AdaptiveAvgPool2d, AdaptiveAvgPool3d]
+    ) -> Tuple[List[int], List[int], List[int]]:
+        """Return parameters for an equivalent AvgPool.
```
Comment (f-dangel) [inline backpack/core/derivatives/adaptive_avg_pool_nd.py:57]: Discussion: Can this naming convention cause confusion? Module parameters could be misinterpreted as the items of `module.parameters()`. Would changing `parameters` to `hyperparameters` resolve such a misunderstanding?

Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:57]: Yes, I agree that it can cause confusion, especially for third party users.
`check_parameters` -> `check_module_parameters`
`get_parameters` -> `get_avg_pool_equivalent_parameters`
I admit that these names are lengthy but I think they avoid any misunderstanding. If misunderstanding is still possible, then `hyperparameters` is better

Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:57]: Do you agree to do the following?
- leave `check_parameters` for now
- change `get_parameters` to `get_avg_pool_equivalent_parameters`

Comment (f-dangel) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:57]: Agree. Linking to https://github.com/fKunstner/backpack-discuss/issues/105

File: test/adaptive_avg_pool/problem.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
+"""Test problems for the AdaptiveAvgPool shape checker."""
+from __future__ import annotations
```
Comment (f-dangel) [inline test/adaptive_avg_pool/problem.py:2]: Need your feedback here: Why is this necessary?

Comment (schaefertim) [reply test/adaptive_avg_pool/problem.py:2]: Because `class AdaptiveAvgPoolProblem` is referenced before definition in a type hint.
Referenced in line 25: `def make_test_problems(settings: List[Dict[str, Any]]) -> List[AdaptiveAvgPoolProblem]:`
Defined in line 82: `class AdaptiveAvgPoolProblem:`

File: backpack/core/derivatives/adaptive_avg_pool_nd.py (hunk 13 above)
Comment (schaefertim) [reply backpack/core/derivatives/adaptive_avg_pool_nd.py:57]: Changed `get_parameters` to `get_avg_pool_parameters`.
Because this method is shared with parent class `AvgPoolNDDerivatives` the name `get_avg_pool_equivalent_parameters` doesn't make sense.

File: backpack/core/derivatives/avgpoolnd.py [hunk 15]
```diff
@@ -31,11 +33,29 @@ def __init__(self, N):
...
+        """
+        return module.stride, module.kernel_size, module.padding
+
     def hessian_is_zero(self):
         return True
 
     def ea_jac_t_mat_jac_prod(self, module, g_inp, g_out, mat):
         """Use fact that average pooling can be implemented as conv."""
+        self.check_parameters(module)
         if self.N == 1:
```
Comment (f-dangel) [inline backpack/core/derivatives/avgpoolnd.py:59]: Refactoring suggestion to avoid the `raise NotImplementedError` (I think it's there because otherwise variables may be undefined below): Compute `C, in_features, out_features, shape_out` independent of `self.N`.
```python
C = module.input0.shape[1]
shape_out = (1,) + tuple(module.input0.shape[2:])
in_features = module.input0.shape[2:].numel()
out_features = module.output.shape[2:].numel()
```

File: backpack/core/derivatives/avgpoolnd.py [hunk 16]
```diff
@@ -151,6 +169,8 @@ def __apply_jacobian_t_of(self, module, mat):
         elif self.N == 3:
```
Comment (f-dangel) [inline backpack/core/derivatives/avgpoolnd.py:169]: Refactoring suggestion to avoid the `NotImplementedError`:
```python
output_size = (V_N_C_in, C_for_conv_t) + tuple(module.input0.shape[2:])
```

File: backpack/core/derivatives/avgpoolnd.py (hunk 15 above)
Comment (schaefertim) [reply backpack/core/derivatives/avgpoolnd.py:59]: Refactored with slightly different version
```
C = module.input0.shape[1]
shape_out = (1,) + tuple(module.output.shape[2:])
in_features = module.input0.shape[1:].numel()
out_features = module.output.shape[1:].numel()
```

File: backpack/core/derivatives/avgpoolnd.py (hunk 16 above)
Comment (schaefertim) [reply backpack/core/derivatives/avgpoolnd.py:169]: Worked out.
